src/parser/jobs/brd/modules/Snapshots.tsx

```diff
@@ -0,0 +1,260 @@
+
+import styles from './Snapshots.module.css'
+
+const SNAPSHOTTERS = [
+	ACTIONS.IRON_JAWS.id,
+	ACTIONS.CAUSTIC_BITE.id,
+	ACTIONS.STORMBITE.id,
+]
+
+const DOT_STATUSES = [
+	STATUSES.CAUSTIC_BITE.id,
+	STATUSES.STORMBITE.id,
+]
+
+const PERSONAL_STATUSES = [
+	STATUSES.MEDICATED.id,
+	STATUSES.RAGING_STRIKES.id,
+]
+
+const BAD_REFRESH_SEVERITY = {
+	1: SEVERITY.MINOR,
+	2: SEVERITY.MEDIUM,
+	3: SEVERITY.MAJOR,
+}
+
+interface Snapshot {
+	// The event that applied the DoT(s).
+	snapEvent: CastEvent
+
+	// All statuses observed at the time of casting the snapshot.
+	statuses: Status[]
+
+	// Indicates whether the snap was an IJ that did not refresh DoTs.
+	badRefresh: boolean
+}
+
+class Target {
+	public name: string
+	public key: string
+	public snapshots: Snapshot[] = []
+	private currentSnapshot?: Snapshot
+	private statuses = new Set<Status>()
+
+	constructor(name: string, key: string) {
+		this.name = name
+		this.key = key
+	}
+
+	public addStatus(status: Status) {
+		this.statuses.add(status)
+	}
+
+	public removeStatus(status: Status) {
+		this.statuses.delete(status)
+	}
+
+	public addSnapshot(snap: CastEvent, playerStatuses: Status[]) {
+		// If IJ didn't refresh a caustic bite, it was a misuse
+		// (IJ on storm bite is technically potency neutral, though odd)
+		const isBadRefresh = (snap.ability.guid === ACTIONS.IRON_JAWS.id)
+							&& !this.statuses.has(STATUSES.CAUSTIC_BITE)
+
+		this.currentSnapshot = {
+			snapEvent: snap,
+			statuses: [...Array.from(this.statuses), ...playerStatuses],
+			badRefresh: isBadRefresh,
+		}
+		this.snapshots.push(this.currentSnapshot)
+	}
+}
+
+export default class Snapshots extends Module {
+	static handle = 'snapshots'
+	static title = t('brd.snapshots.title')`Snapshots`
+	static debug = false
+
+	@dependency private combatants!: Combatants
+	@dependency private data!: Data
+	@dependency private suggestions!: Suggestions
+	@dependency private util!: Util
+
+	private targets: Map<string, Target> = new Map()
+
+	protected init() {
+		this.addEventHook('cast', {by: 'player', abilityId: SNAPSHOTTERS}, this.onSnapshot)
+		this.addEventHook(['applydebuff', 'refreshdebuff'], this.onApply)
+		this.addEventHook('removedebuff', this.onRemove)
+		this.addEventHook('complete', this.onComplete)
+	}
+
+	private onApply(event: AbilityEvent) {
+		if (event.targetIsFriendly) { return }
+
+		const target = this.getTarget(event)
+		const status = this.data.getStatus(event.ability.guid)
+		if (status) {
+			target.addStatus(status)
+		}
+	}
+
+	private onRemove(event: AbilityEvent) {
+		if (event.targetIsFriendly) { return }
+
+		const target = this.getTarget(event)
+		const status = this.data.getStatus(event.ability.guid)
+		if (status) {
+			target.removeStatus(status)
+		}
+	}
+
+	private onSnapshot(event: CastEvent) {
+		const target = this.getTarget(event)
+		target.addSnapshot(event, this.playerStatuses)
+	}
+
+	private getTarget(fields: {targetID?: number, targetInstance?: number}): Target {
```

Comment:
Do these fields not exist on BuffEvent? I can't recall what came out of our Discord chat but I feel like the `onApply`/`onRemove` handlers should be using BuffEvent rather than AbilityEvent, and then this entire function could be replaced with a getter in the Target class, along with what Az suggested.

Reply:
I changed all AbilityEvent types to BuffEvent, I'm not sure what you mean about a getter in the Target class though. Do you mean a static method in the Target class that takes the map as an arg?

Reply:
Basically yeah, but in retrospect it's a bad idea.